### PR TITLE
Merge from micropython v1.22.2

### PR DESCRIPTION
**Pull request by dhalbert**
Merging from `micropython-v1.22-release` branch.

- added background tasks and interrupt checks to new routines `mp_event_wait_ms() `and `mp_event_wait_indefinite()`.
- Note that `QDEF` is now `QDEF0` and `QDEF1`.
- `pyexec.c` is still hard to merge due to `atexit` conditionals. Maybe there's some way of formatting them so there are fewer changes.
- `raspberrypi/libm_extra.c `fixes ROM floats to handle infinities correctly, but I did not pick this up yet.

As of the initial draft PR, at least one board from each port has been compiled. `atmel-samd`, `nordic,`espressif`, `rasbperrypi` have had some basic testing done, blink for most, BLE for nordic, and wifi for `espressif`. I'm submitting the PR to see how the tests and doc builds work out.

## Update: out of draft mode.

Everything builds except for two too-large builds, which I will fix soon, either here or in a later PR. Smoke tests work. 

- I added quite a few more `CIRCUITPY-CHANGE` annotations. They are more or less complete now.
- I updated the https://github.com/adafruit/Adafruit_CircuitPython_asyncio library. The changes are currently in a branch: https://github.com/adafruit/Adafruit_CircuitPython_asyncio/tree/v1.22-merge. That branch is what is selected for `frozen/Adafruit_CircuitPython_asyncio` at the moment. I need to test it against 9.1.1 and see if the changes still work for 9.1.1 before I merge it into main in the library. So anyone using `asyncio` with this build should use that branch for now.

### CODE REVIEW

**Comment by dhalbert** [comment]
Ready for review. There are two builds failing due to space issues. This is a bit surprising, but I don't think this should hold up reviewing the whole thing. I'm thinking there may be some extra math code or something like that. I will look and see what grew.

**Comment by dhalbert** [comment]
[also put in OP]

## Update: out of draft mode.

Everything builds except for two too-large builds, which I will fix soon, either here or in a later PR. Smoke tests work. 

- I added quite a few more `CIRCUITPY-CHANGE` annotations. They are more or less complete now.
- I updated the https://github.com/adafruit/Adafruit_CircuitPython_asyncio library. The changes are currently in a branch: https://github.com/adafruit/Adafruit_CircuitPython_asyncio/tree/v1.22-merge. That branch is what is selected for `frozen/Adafruit_CircuitPython_asyncio` at the moment. I need to test it against 9.1.1 and see if the changes still work for 9.1.1 before I merge it into main in the library. So anyone using `asyncio` with this build should use that branch for now.

**Comment by dhalbert** [comment]
@tannewt all green now

File: py/objint.c [hunk 1]
```diff
@@ -528,6 +530,7 @@ STATIC mp_obj_t int_from_bytes(size_t n_args, const mp_obj_t *pos_args, mp_map_t
 STATIC MP_DEFINE_CONST_FUN_OBJ_KW(int_from_bytes_fun_obj, 3, int_from_bytes);
 STATIC MP_DEFINE_CONST_CLASSMETHOD_OBJ(int_from_bytes_obj, MP_ROM_PTR(&int_from_bytes_fun_obj));
 
+// CIRCUITPY-CHANGE: supprts signed
```
Comment (tannewt) [inline py/objint.c:533]:
```suggestion
// CIRCUITPY-CHANGE: supports signed
```

File: py/qstr.h [hunk 2]
```diff
@@ -97,7 +112,9 @@ void qstr_pool_info(size_t *n_pool, size_t *n_qstr, size_t *n_str_data_bytes, si
 void qstr_dump_data(void);
 
 #if MICROPY_ROM_TEXT_COMPRESSION
-void mp_decompress_rom_string(byte *dst, mp_rom_error_text_t src);
+// CIRCUITPY-CHANGE: TODO not const previusly, should it be??
+// TEST COMPILE with const ******************
```
Comment (tannewt) [inline py/qstr.h:116]: Want to do this?

File: shared-bindings/rainbowio/__init__.h [hunk 3]
```diff
@@ -6,6 +6,9 @@
 
 #pragma once
 #include <stdint.h>
-#include "py/misc.h"
+
+#include "py/obj.h"
+
+///#include "shared-module/rainbowio/__init__.h"
```
Comment (tannewt) [inline shared-bindings/rainbowio/__init__.h:12]:
```suggestion
```

File: tools/merge_micropython.py [hunk 4]
```diff
@@ -47,6 +46,8 @@
     except sh.ErrorReturnCode_128:
         pass
 
+sys.exit(0)
+
```
Comment (tannewt) [inline tools/merge_micropython.py:50]: Want to commit this?

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for doing this! Just a few minor things. Mostly typos

File: py/mkrules.mk [hunk 5]
```diff
@@ -158,6 +168,7 @@ $(HEADER_BUILD)/compressed.collected: $(HEADER_BUILD)/compressed.split
 # will be created if they don't exist.
 OBJ_DIRS = $(sort $(dir $(OBJ)))
 $(OBJ): | $(OBJ_DIRS)
+// CIRCUITPY-CHANGE: use $(Q)
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply py/mkrules.mk:171]: It is weird this didn't cause an error!

File: py/qstr.h (hunk 2 above)
Comment (dhalbert) [reply py/qstr.h:116]: I did :), and forgot to remove the comment.

File: tools/merge_micropython.py (hunk 4 above)
Comment (dhalbert) [reply tools/merge_micropython.py:50]: Oops. I run pieces of this at a time.

**Review by tannewt — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**